x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/assignment/TrainedModelAssignmentNodeService.java

```diff
@@ -241,22 +228,38 @@ void loadQueuedModels() {
         loadingModels.addAll(loadingToRetry);
     }
 
+    public void gracefullyStopDeploymentAndNotify(
+        TrainedModelDeploymentTask task,
+        String reason,
+        ActionListener<AcknowledgedResponse> listener
+    ) {
+        logger.debug(() -> format("[%s] Gracefully stopping deployment due to reason %s", task.getDeploymentId(), reason));
+
+        stopAndNotifyHelper(task, reason, listener, deploymentManager::stopAfterCompletingPendingWork);
+    }
+
     public void stopDeploymentAndNotify(TrainedModelDeploymentTask task, String reason, ActionListener<AcknowledgedResponse> listener) {
-        final RoutingInfoUpdate updateToStopped = RoutingInfoUpdate.updateStateAndReason(
-            new RoutingStateAndReason(RoutingState.STOPPED, reason)
-        );
+        logger.debug(() -> format("[%s] Forcefully stopping deployment due to reason %s", task.getDeploymentId(), reason));
+
+        stopAndNotifyHelper(task, reason, listener, deploymentManager::stopDeployment);
+    }
+
+    private void stopAndNotifyHelper(
+        TrainedModelDeploymentTask task,
+        String reason,
+        ActionListener<AcknowledgedResponse> listener,
+        Consumer<TrainedModelDeploymentTask> stopDeploymentFunc
+    ) {
+        // Removing the entry from the map to avoid the possibility of a node shutdown triggering a concurrent graceful stopping of the
+        // process while we are attempting to forcefully stop the native process
+        // The graceful stopping will only occur if there is an entry in the map
+        deploymentIdToTask.remove(task.getDeploymentId());
```

Comment:
I think adding this caused a bug where the test `testStoppingDeploymentShouldTriggerRebalance` stop deployment call ended up causing a model to be loaded because the check was looking for the absence of a task within `deploymentIdToTask`. This resulted in a task that was never cancelled or unregistered which blocks the MachineLearning `_feature/_reset` handler because it waits for all tasks to complete. I think the fix is to ensure that the assignment is not stopping before adding a model to be loaded.

<details><summary>Bug logs</summary>

```
[2023-08-22T21:51:40,665][DEBUG][o.e.x.m.a.TransportStopTrainedModelDeploymentAction] [javaRestTest-1] [stopping_triggers_rebalance_1] Received request to undeploy
[2023-08-22T21:51:40,667][DEBUG][o.e.x.m.i.a.TrainedModelAssignmentClusterService] [javaRestTest-1] updated assignments: {"stopping_triggers_rebalance_2":{"task_parameters":{"model_id":"stopping_triggers_rebalance_2","deployment_id":"stopping_triggers_rebalance_2","model_bytes":1630,"threads_per_allocation":1,"number_of_allocations":1,"queue_capacity":1024,"cache_size":"1630b","priority":"normal","per_deployment_memory_bytes":0,"per_allocation_memory_bytes":0},"routing_table":{},"assignment_state":"starting","reason":"Could not assign (more) allocations on node [BhBlT0rIRKOqmiBp4XnIFw]. Reason: This node has insufficient allocated processors. Available processors [32], free processors [0], processors required for each allocation of this model [1]|Could not assign (more) allocations on node [CIrVFGyRQDCFp-KN6xG1Og]. Reason: This node has insufficient allocated processors. Available processors [32], free processors [0], processors required for each allocation of this model [1]","start_time":"2023-08-22T21:51:40.613514595Z","max_assigned_allocations":0},"stopping_triggers_rebalance_1":{"task_parameters":{"model_id":"stopping_triggers_rebalance_1","deployment_id":"stopping_triggers_rebalance_1","model_bytes":1630,"threads_per_allocation":1,"number_of_allocations":100,"queue_capacity":1024,"cache_size":"1630b","priority":"normal","per_deployment_memory_bytes":0,"per_allocation_memory_bytes":0},"routing_table":{"BhBlT0rIRKOqmiBp4XnIFw":{"current_allocations":32,"target_allocations":32,"routing_state":"started","reason":""},"CIrVFGyRQDCFp-KN6xG1Og":{"current_allocations":32,"target_allocations":32,"routing_state":"starting","reason":""}},"assignment_state":"stopping","reason":"client API call","start_time":"2023-08-22T21:51:39.837120921Z","max_assigned_allocations":64}}
[2023-08-22T21:51:40,684][DEBUG][o.e.x.m.i.a.TrainedModelAssignmentClusterService] [javaRestTest-1] [stopping_triggers_rebalance_1] updating routing table entry for node [BhBlT0rIRKOqmiBp4XnIFw], update [RoutingInfoUpdate{numberOfAllocations=Optional.empty, stateAndReason=Optional[RoutingStateAndReason{reason='undeploy_trained_model (api)', state=stopping}]}]
[2023-08-22T21:51:40,685][DEBUG][o.e.x.m.i.a.TrainedModelAssignmentClusterService] [javaRestTest-1] [stopping_triggers_rebalance_1] updating routing table entry for node [CIrVFGyRQDCFp-KN6xG1Og], update [RoutingInfoUpdate{numberOfAllocations=Optional.empty, stateAndReason=Optional[RoutingStateAndReason{reason='undeploy_trained_model (api)', state=stopping}]}]
[2023-08-22T21:51:40,686][DEBUG][o.e.x.m.i.a.TrainedModelAssignmentClusterService] [javaRestTest-1] [stopping_triggers_rebalance_1] requested update from node [BhBlT0rIRKOqmiBp4XnIFw] while stopping; update was [RoutingInfoUpdate{numberOfAllocations=Optional.empty, stateAndReason=Optional[RoutingStateAndReason{reason='undeploy_trained_model (api)', state=stopping}]}]
[2023-08-22T21:51:40,686][DEBUG][o.e.x.m.i.a.TrainedModelAssignmentClusterService] [javaRestTest-1] [stopping_triggers_rebalance_1] requested update from node [CIrVFGyRQDCFp-KN6xG1Og] while stopping; update was [RoutingInfoUpdate{numberOfAllocations=Optional.empty, stateAndReason=Optional[RoutingStateAndReason{reason='undeploy_trained_model (api)', state=stopping}]}]
[2023-08-22T21:51:40,686][DEBUG][o.e.x.m.i.a.TrainedModelAssignmentNodeService] [javaRestTest-1] [stopping_triggers_rebalance_1] deployment routing info was updated with [RoutingInfoUpdate{numberOfAllocations=Optional.empty, stateAndReason=Optional[RoutingStateAndReason{reason='undeploy_trained_model (api)', state=stopping}]}] and master notified
[2023-08-22T21:51:40,686][DEBUG][o.e.x.m.i.d.TrainedModelDeploymentTask] [javaRestTest-1] [stopping_triggers_rebalance_1] Stopping due to reason [undeploy_trained_model (api)]
[2023-08-22T21:51:40,687][INFO ][o.e.x.m.i.d.DeploymentManager] [javaRestTest-1] [stopping_triggers_rebalance_1] Stopping deployment, reason [undeploy_trained_model (api)]
[2023-08-22T21:51:40,697][DEBUG][o.e.x.m.i.a.TrainedModelAssignmentClusterService] [javaRestTest-1] [stopping_triggers_rebalance_1] updating routing table entry for node [BhBlT0rIRKOqmiBp4XnIFw], update [RoutingInfoUpdate{numberOfAllocations=Optional.empty, stateAndReason=Optional[RoutingStateAndReason{reason='undeploy_trained_model (api)', state=stopped}]}]
[2023-08-22T21:51:40,697][DEBUG][o.e.x.m.i.a.TrainedModelAssignmentClusterService] [javaRestTest-1] updated assignments: {"stopping_triggers_rebalance_2":{"task_parameters":{"model_id":"stopping_triggers_rebalance_2","deployment_id":"stopping_triggers_rebalance_2","model_bytes":1630,"threads_per_allocation":1,"number_of_allocations":1,"queue_capacity":1024,"cache_size":"1630b","priority":"normal","per_deployment_memory_bytes":0,"per_allocation_memory_bytes":0},"routing_table":{},"assignment_state":"starting","reason":"Could not assign (more) allocations on node [BhBlT0rIRKOqmiBp4XnIFw]. Reason: This node has insufficient allocated processors. Available processors [32], free processors [0], processors required for each allocation of this model [1]|Could not assign (more) allocations on node [CIrVFGyRQDCFp-KN6xG1Og]. Reason: This node has insufficient allocated processors. Available processors [32], free processors [0], processors required for each allocation of this model [1]","start_time":"2023-08-22T21:51:40.613514595Z","max_assigned_allocations":0},"stopping_triggers_rebalance_1":{"task_parameters":{"model_id":"stopping_triggers_rebalance_1","deployment_id":"stopping_triggers_rebalance_1","model_bytes":1630,"threads_per_allocation":1,"number_of_allocations":100,"queue_capacity":1024,"cache_size":"1630b","priority":"normal","per_deployment_memory_bytes":0,"per_allocation_memory_bytes":0},"routing_table":{"CIrVFGyRQDCFp-KN6xG1Og":{"current_allocations":32,"target_allocations":32,"routing_state":"starting","reason":""}},"assignment_state":"stopping","reason":"client API call","start_time":"2023-08-22T21:51:39.837120921Z","max_assigned_allocations":64}}
[2023-08-22T21:51:40,711][DEBUG][o.e.x.m.i.a.TrainedModelAssignmentNodeService] [javaRestTest-1] [stopping_triggers_rebalance_1] preparing to load model [stopping_triggers_rebalance_1] with task params: {"model_id":"stopping_triggers_rebalance_1","deployment_id":"stopping_triggers_rebalance_1","model_bytes":1630,"threads_per_allocation":1,"number_of_allocations":32,"queue_capacity":1024,"cache_size":"1630b","priority":"normal","per_deployment_memory_bytes":0,"per_allocation_memory_bytes":0}
[2023-08-22T21:51:40,732][DEBUG][o.e.x.m.p.l.CppLogMessageHandler] [javaRestTest-1] [stopping_triggers_rebalance_1] [pytorch_inference/154905] [CLogger.cc@338] Logger is logging to named pipe /dev/shm/elastic+elasticsearch+pull-request+part-3/x-pack/plugin/ml/qa/native-multi-node-tests/build/testclusters/javaRestTest-1/tmp/pytorch_inference_stopping_triggers_rebalance_1_log_124237
[2023-08-22T21:51:40,732][DEBUG][o.e.x.m.p.l.CppLogMessageHandler] [javaRestTest-1] [stopping_triggers_rebalance_1] [pytorch_inference/154905] [Main.cc@227] pytorch_inference (64 bit): Version 8.11.0-SNAPSHOT (Build afc7449dbf3d26) Copyright (c) 2023 Elasticsearch BV
[2023-08-22T21:51:40,732][DEBUG][o.e.x.m.p.l.CppLogMessageHandler] [javaRestTest-1] [stopping_triggers_rebalance_1] [pytorch_inference/154905] [CProcessPriority_Linux.cc@38] Successfully increased OOM killer adjustment via /proc/self/oom_score_adj
[2023-08-22T21:51:40,732][DEBUG][o.e.x.m.p.l.CppLogMessageHandler] [javaRestTest-1] [stopping_triggers_rebalance_1] [pytorch_inference/154905] [CSystemCallFilter_Linux.cc@162] Seccomp BPF filters available
[2023-08-22T21:51:40,732][DEBUG][o.e.x.m.p.l.CppLogMessageHandler] [javaRestTest-1] [stopping_triggers_rebalance_1] [pytorch_inference/154905] [CSystemCallFilter_Linux.cc@188] Seccomp BPF installed
[2023-08-22T21:51:40,793][DEBUG][o.e.x.m.i.d.DeploymentManager] [javaRestTest-1] [stopping_triggers_rebalance_1] process started
[2023-08-22T21:51:40,802][DEBUG][o.e.x.m.i.d.DeploymentManager] [javaRestTest-1] [stopping_triggers_rebalance_1] Forcibly stopping process
[2023-08-22T21:51:40,810][DEBUG][o.e.x.m.p.l.CppLogMessageHandler] [javaRestTest-1] [controller/124362] [CResponseJsonWriter.cc@42] Wrote controller response - id: 127 success: true reason: Process with PID 154905 killed
[2023-08-22T21:51:40,811][DEBUG][o.e.x.m.i.p.p.PyTorchStateStreamer] [javaRestTest-1] model [stopping_triggers_rebalance_1] state restored in [0] documents from index [.ml-inference-native-000001]
[2023-08-22T21:51:40,811][INFO ][o.e.x.m.i.p.p.PyTorchStateStreamer] [javaRestTest-1] [stopping_triggers_rebalance_1] loading model state cancelled
[2023-08-22T21:51:40,812][DEBUG][o.e.x.m.i.d.DeploymentManager] [javaRestTest-1] [stopping_triggers_rebalance_1] model loaded but process is stopped
[2023-08-22T21:51:40,812][DEBUG][o.e.x.m.p.l.CppLogMessageHandler] [javaRestTest-1] [controller/124362] [CResponseJsonWriter.cc@42] Wrote controller response - id: 128 success: true reason: Process with PID 154905 killed
[2023-08-22T21:51:40,817][DEBUG][o.e.x.m.i.a.TrainedModelAssignmentNodeService] [javaRestTest-1] [stopping_triggers_rebalance_1] Updating routing state to stopped
[2023-08-22T21:51:40,817][DEBUG][o.e.x.m.i.a.TrainedModelAssignmentClusterService] [javaRestTest-1] [stopping_triggers_rebalance_1] updating routing table entry for node [CIrVFGyRQDCFp-KN6xG1Og], update [RoutingInfoUpdate{numberOfAllocations=Optional.empty, stateAndReason=Optional[RoutingStateAndReason{reason='undeploy_trained_model (api)', state=stopped}]}]
[2023-08-22T21:51:40,817][WARN ][o.e.x.m.i.a.TrainedModelAssignmentNodeService] [javaRestTest-1] [stopping_triggers_rebalance_1] Start deployment failed java.lang.IllegalStateException: model loaded but process is stopped
```

</details>